File: app/src/main/java/com/blues/gankio/old/model_android/view/AndroidFragment.java
package com.blues.gankio.old.model_android.view;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.LinearLayout;

import com.blues.LoadingDialog;
import com.blues.R;
import com.blues.framework.base.BaseFragment;
import com.blues.gankio.old.model_android.AndroidContract;
import com.blues.gankio.old.model_android.model.AndroidEntity;
import com.blues.gankio.old.model_android.presenter.AndroidPresenter;
import com.jude.easyrecyclerview.EasyRecyclerView;
import com.jude.easyrecyclerview.adapter.BaseViewHolder;
import com.jude.easyrecyclerview.adapter.RecyclerArrayAdapter;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.header.ClassicsHeader;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;

public class AndroidFragment extends BaseFragment implements AndroidContract.iContractView, OnRefreshListener {

    private SmartRefreshLayout androidSr;
    private ViewStub networkVS;
    private View v;
    private AndroidContract.iContractPresenter iPresenter;
    private RecyclerArrayAdapter<AndroidEntity.ResultsBean> mAdapter;
    private LoadingDialog mLoading;

    @Override
    public int setLayoutResourceId() {
        return R.layout.fragment_android;
    }

    @Override
    public void initLayout(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        EasyRecyclerView androidErv = rootView.findViewById(R.id.rv_android);
        androidSr = rootView.findViewById(R.id.sr_android);
        networkVS = rootView.findViewById(R.id.view_stub_network);
        //设置 Header 为 经典 样式 带最后刷新时间
        androidSr.setRefreshHeader(new ClassicsHeader(getmContext()).setEnableLastTime(true));
        androidSr.setEnableHeaderTranslationContent(true);

        mPresenter = new AndroidPresenter(this);

        //初始化数据
        iPresenter.initData();
        androidErv.setLayoutManager(new LinearLayoutManager(getmContext()));
        androidErv.setAdapter(mAdapter = new RecyclerArrayAdapter<AndroidEntity.ResultsBean>(getmContext()) {
            @Override
            public BaseViewHolder OnCreateViewHolder(ViewGroup parent, int viewType) {
                return new AndroidViewHolder(parent);
            }
        });
        mAdapter.setMore(R.layout.view_loading_more, new RecyclerArrayAdapter.OnMoreListener() {
            @Override
            public void onMoreShow() {
                iPresenter.loadMore();
            }

            @Override
            public void onMoreClick() {

            }
        });

        mAdapter.setNoMore(R.layout.view_load_no_more);
    }

    @Override
    public void lazyFetchData() {

    }

    @Override
    public void setListener() {
        androidSr.setEnableRefresh(true);
        androidSr.setOnRefreshListener(this);
    }

    @Override
    public void setData(List<AndroidEntity.ResultsBean> result) {
        if (v != null) {
            v.setVisibility(View.GONE);
        }
        for (AndroidEntity.ResultsBean data : result) {
            mAdapter.add(data);
        }
        mAdapter.notifyDataSetChanged();
    }

    @Override
    public void setPresenter(AndroidContract.iContractPresenter presenter) {
        this.iPresenter = presenter;
    }

    @Override
    public void showLoading() {
        mLoading = new LoadingDialog(getmContext());
        mLoading.show();
    }

    @Override
    public void showFinished() {
        if (mLoading != null) {
            mLoading.dismiss();
        }
        androidSr.finishRefresh();
    }

    @Override
    public void showError() {
        try {
            v = networkVS.inflate();
            LinearLayout blankLl = v.findViewById(R.id.error_blank);
            blankLl.setOnClickListener(v1 -> {
                mAdapter.clear();
                iPresenter.initData();
            });
        } catch (Exception e) {
            e.printStackTrace();
            v.setVisibility(View.VISIBLE);
        }
        if (mLoading != null) {
            mLoading.dismiss();
        }
    }

    @Override
    public void showEmpty() {
        if (mLoading != null) {
            mLoading.dismiss();
        }
    }

    @Override
    public void onRefresh(@NonNull RefreshLayout refreshLayout) {
        mAdapter.clear();
        iPresenter.initData();
    }
}

File: app/src/main/java/com/blues/gankio/v1/view/AndroidFragment.kt
package com.blues.gankio.v1.view

import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.jude.easyrecyclerview.adapter.RecyclerArrayAdapter
import com.blues.R
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import com.jude.easyrecyclerview.EasyRecyclerView
import com.scwang.smartrefresh.layout.header.ClassicsHeader
import androidx.recyclerview.widget.LinearLayoutManager
import com.jude.easyrecyclerview.adapter.RecyclerArrayAdapter.OnMoreListener
import com.blues.framework.base.BaseKoinFragment
import com.blues.gankio.v1.api.GankType
import com.blues.gankio.v1.model.GankBean
import com.blues.gankio.v1.vm.GankViewModel
import com.jude.easyrecyclerview.adapter.BaseViewHolder
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.listener.OnRefreshListener
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.viewModel

class AndroidFragment : BaseKoinFragment(), OnRefreshListener {

    private val androidViewModel: GankViewModel by viewModel()

    private lateinit var androidSr: SmartRefreshLayout
    private lateinit var mAdapter: RecyclerArrayAdapter<GankBean.ResultsBean>

    override fun collect() {
        lifecycleScope.launch {
            androidViewModel.listData.collect {
                it.results.let { list ->
                    mAdapter.apply {
                        addAll(list)
                        notifyDataSetChanged()
                    }
                }
            }
        }
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
        mAdapter.clear()
        androidViewModel.initData(GankType.ANDROID.type)
    }

    override fun getLayoutId(): Int = R.layout.fragment_android

    override fun initData(inflater: LayoutInflater, container: ViewGroup?, saveInstanced: Bundle?) {

        //初始化数据
        androidViewModel.initData(GankType.ANDROID.type)

        rootView.findViewById<EasyRecyclerView>(R.id.rv_android).apply {
            setLayoutManager(LinearLayoutManager(requireContext()))
            adapter = object : RecyclerArrayAdapter<GankBean.ResultsBean>(requireContext()) {
                override fun OnCreateViewHolder(
                    parent: ViewGroup,
                    viewType: Int
                ): BaseViewHolder<*> {
                    return AndroidViewHolder(parent)
                }
            }.also { mAdapter = it }
        }
        androidSr = rootView.findViewById<SmartRefreshLayout>(R.id.sr_android)
            .apply { //设置 Header 为 经典 样式 带最后刷新时间
                setRefreshHeader(ClassicsHeader(requireContext()).setEnableLastTime(true))
                setEnableHeaderTranslationContent(true)
                setEnableRefresh(true)
                setOnRefreshListener(this@AndroidFragment)
            }

        mAdapter.apply {
            setNoMore(R.layout.view_load_no_more)
            setMore(R.layout.view_loading_more, object : OnMoreListener {
                override fun onMoreShow() {
                    androidViewModel.loadMore(GankType.ANDROID.type)
                }

                override fun onMoreClick() {}
            })
        }
    }
}